### PR TITLE
Add OCM backup label to hub config map

### PR DESCRIPTION
**Pull request by ELENAGER**
OCM backup label ensures that an OCM backup picks up hub config map to backup and restore for hub recovery situations.

Fixes: #375

### CODE REVIEW

File: controllers/drcluster_controller_test.go [hunk 1]
```diff
@@ -438,6 +438,9 @@ var _ = Describe("DRClusterController", func() {
 				By("changing the S3Profile in ramen config to an invalid value")
 				newS3Profiles := s3Profiles[0:]
 				s3Profiles[5].S3Bucket = bucketNameFail
+				Expect(k8sClient.Get(context.TODO(),
+					types.NamespacedName{Name: controllers.HubOperatorConfigMapName, Namespace: ramenNamespace},
+					configMap)).To(Succeed())
```
Comment (nirs) [inline controllers/drcluster_controller_test.go:443]: Not clear what this line tests. If we cannot use more clear code maybe add a comment?

File: controllers/drpolicy_controller.go [hunk 2]
```diff
@@ -382,6 +382,16 @@ func (r *DRPolicyReconciler) configMapMapFunc(configMap client.Object) []reconci
...
 
+	labelAdded := util.AddLabel(configMap, util.OCMBackupLabelKey, util.OCMBackupLabelValue)
+
+	if labelAdded {
+		if err := r.Update(context.TODO(), configMap); err != nil {
+			r.Log.Error(err, "Failed to add OCM backup label to ramen-hub-operator-config map")
+
+			return []reconcile.Request{}
+		}
+	}
```
Comment (nirs) [inline controllers/drpolicy_controller.go:393]: Why add label dynamically? Can we install a config map with a label when deploying?

Also when this code run? The commit message should describe the solution - why we add label dynamically on which events we add the label.

File: controllers/drpolicy_controller.go [hunk 3]
```diff
@@ -382,6 +382,16 @@ func (r *DRPolicyReconciler) configMapMapFunc(configMap client.Object) []reconci
 		return []reconcile.Request{}
 	}
 
+	labelAdded := util.AddLabel(configMap, util.OCMBackupLabelKey, util.OCMBackupLabelValue)
```
Comment (ShyamsundarR) [inline controllers/drpolicy_controller.go:385]: I would move this to [ConfigMapGet](https://github.com/RamenDR/ramen/blob/dbb1fe057731b32095a642ff8891297d559e8589/controllers/ramenconfig.go#L207) function instead.

This routine is called on every config map update as part of the reconciler watch. Instead, moving this to the routine above helps add the label across the code base where the config map is accessed, on the hub as well as on the managed cluster (although on the managed cluster there is no backup initiated, so maybe we can latch the label addition out).

File: controllers/drcluster_controller_test.go (hunk 1 above)
Comment (ELENAGER) [reply controllers/drcluster_controller_test.go:443]: @nirs , you are right, it is not clear. The next line (s3ProfilesStore(newS3Profiles)) is changing ConfigMap. Before doing it, it is necessary to get the latest ConfigMap (k8sClient.Get is reloading it). 

We didn't need it before adding label in this PR. Nobody changed ConfigMap, only read from it.

File: controllers/drpolicy_controller.go (hunk 2 above)
Comment (ELENAGER) [reply controllers/drpolicy_controller.go:393]: @nirs yes, I also wanted to add the label only once upon initialization. But I didn't find the right place, so I consulted with @BenamarMk and he suggested to do it inside the update. Maybe I didn't understand him right, let's wait for his review.

Comment (BenamarMk) [reply controllers/drpolicy_controller.go:393]: You absolutely can create it statically @ELENAGER

File: controllers/drcluster_controller_test.go (hunk 1 above)
Comment (ELENAGER) [reply controllers/drcluster_controller_test.go]: In order to make code clearer, I moved the reloading of the ConfigMap into rameConfig_test file.

File: controllers/drpolicy_controller.go (hunk 2 above)
Comment (ELENAGER) [reply controllers/drpolicy_controller.go:393]: I added the static label creatiion. Now ConfigMap will be initialized with OCM backup label.  Only in case, that user changed ConfigMap, we will ensure this label exists.

**Review by nirs — COMMENTED** [review]
It will be helpful to separate the static label addition from the update. The static change is simple and looks ready so we can merge it now, the other part I'm not sure.

Even if you don't care about making progress faster, each commit should do one logical change. and here we have 3 changes:

- Add backup label to the default config map
- Add backup label to ramenctl config map
- Ensure that backup label exist on the hub

File: controllers/ramenconfig_test.go [hunk 4]
```diff
@@ -9,16 +9,33 @@ import (
...
 	Expect(err).To(Succeed())
 
-	configMap.Data[controllers.ConfigMapRamenConfigKeyName] = string(ramenConfigYaml)
+	retryErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var err error
 
-	Expect(k8sClient.Update(context.TODO(), configMap)).To(Succeed())
+		err = k8sClient.Get(context.TODO(),
+			types.NamespacedName{Name: controllers.HubOperatorConfigMapName, Namespace: ramenNamespace},
+			configMap)
```
Comment (nirs) [inline controllers/ramenconfig_test.go:26]: This is accepted by the linter but is unreadable. When creating structs always
use one key: value pair per line style, unless the entire line is very short:

    types.NamespacedName{
        Name: controllers.HubOperatorConfigMapName,
        Namespace: ramenNamespace,
    },

Also this is better done once outside of the retry closure:

    configMapName := types.NamespacedName{
        Name: controllers.HubOperatorConfigMapName,
        Namespace: ramenNamespace,
    }

    retryErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
        err := k8sClient.Get(context.TODO(), configMapName, configMap)
        ...
    })

File: controllers/ramenconfig_test.go [hunk 5]
```diff
@@ -9,16 +9,33 @@ import (
...
+		if err != nil {
+			return err
+		}
+
+		configMap.Data[controllers.ConfigMapRamenConfigKeyName] = string(ramenConfigYaml)
+
+		err = k8sClient.Update(context.TODO(), configMap)
+
+		return err
```
Comment (nirs) [inline controllers/ramenconfig_test.go:35]: Any reason to keep the return value of k8sClient.Update() in err? We can simplify this
to:

    return k8sClient.Update(context.TODO(), configMap)

File: controllers/ramenconfig_test.go (hunk 4 above)
Comment (ELENAGER) [reply controllers/ramenconfig_test.go]: Done

File: controllers/ramenconfig_test.go (hunk 5 above)
Comment (ELENAGER) [reply controllers/ramenconfig_test.go]: Done

File: controllers/ramenconfig_test.go [hunk 6]
```diff
@@ -9,16 +9,32 @@ import (
...
 	"github.com/ramendr/ramen/controllers"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
 	"sigs.k8s.io/yaml"
 )
 
 func configMapUpdate() {
 	ramenConfigYaml, err := yaml.Marshal(ramenConfig)
-	Expect(err).To(Succeed())
+	Expect(err).NotTo(HaveOccurred())
```
Comment (nirs) [inline controllers/ramenconfig_test.go:19]: This is change is not really related, next time better split such changes to another commit (unify error handling).

**Review by nirs — COMMENTED** [review]
Looks good, but the second commit should be squashed into the first.

If you want to split this to multiple commits splitting the static yaml changes from the dynamic updated makes sense.

File: controllers/ramenconfig_test.go (hunk 6 above)
Comment (ELENAGER) [reply controllers/ramenconfig_test.go:19]: Next time I'll prepare several commits

**Review by nirs — APPROVED** [review]

File: controllers/drpolicy_controller.go (hunk 3 above)
Comment (ShyamsundarR) [reply controllers/drpolicy_controller.go:385]: This is fine. Even if we do not remove ConfigMapGet, the label check is on the object passed in and also the predicate is a resource version change predicate, and would only occur if this was changed. As a result calling this often would not occur.
